Fix the unpredictable order randomization issue with randomized content blocks

lgp171188 opened the pull request:
This PR contains changes to the Randomized Content Block XBlock. The XBlock has unpredictable randomization of the order of the selected child blocks due to the usage of sets, which are unordered, for storing the selected blocks. This becomes apparent when all the available child blocks in a library are chosen for a Randomized Content Block to randomize just the order of the child blocks. The fix in this PR modifies the XBlock to store the selected child blocks in a list after randomly shuffling them.

**JIRA tickets**:  TBD - ticket will be created on creating this PR.

**Dependencies**: None

**Sandbox URL**: 
LMS - https://pr18675.sandbox.opencraft.hosting/
Studio - https://studio-pr18675.sandbox.opencraft.hosting/

**Merge deadline**: None

**Testing instructions**:

1. Login to the CMS as a user who can create and edit courses.
1. Create a library with a few child components.
1. Add a unit to a course.
1. Add a Randomized Content Block to the unit.
1. Select the library created above as the source for this XBlock.
1. Select all types of children to be shown.
1. Specify the total number of components present in the library as the number of components to display.
1. Save the unit and publish it.
1. Register a few students to this course.
1. Navigate to the unit containing this Randomized Content Block.
1. Verify that the order of the content blocks are randomized for every student.

**Reviewers**
- [ ] @kaizoku 
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @lgp171188! I've created [OSPR-2539](https://openedx.atlassian.net/browse/OSPR-2539) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).



If you like, you can add yourself to the [AUTHORS](https://github.com/edx/edx-platform/blob/master/AUTHORS)      file for this repo, though that isn't required.     Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst)     file for more information.

lgp171188 commented:
Marking this PR as WIP since some changes are needed to fix the failing tests.

mduboseedx commented:
@lgp171188 Thanks for the contribution. Feel free to ping me once the tests pass.

lgp171188 commented:
jenkins run bokchoy

xitij2000 commented:
jenkins run bokchoy

kaizoku commented:
Looking and working well @lgp171188. Thanks for updating this.
:+1:

- [✓] I tested this: (deployed in devstack, checked functionality and randomization of question order between user accounts, ensured that it persists order for each user)
- [✓] I read through the code
- [n/a] I checked for accessibility issues
- [n/a] Includes documentation

lgp171188 commented:
@mduboseedx, @kaizoku, all the checks pass now and this is now ready for review.

mduboseedx commented:
@edx/educator-all Could this be given a review? It is a change to the Randomized Content Block

kaizoku commented:
Thanks @lgp171188 , this is working well for me.
:+1:

- [✓] I tested this: (deployed in devstack, setup randomized content block, checked question order for different users)
- [✓] I read through the code
- [n/a] I checked for accessibility issues
- [n/a] Includes documentation

lgp171188 commented on common/lib/xmodule/xmodule/library_content_module.py:
```diff
@@ -293,19 +293,15 @@ def publish_selected_children_events(cls, block_keys, format_block_keys, publish
         """
-        if hasattr(self, "_selected_set"):
-            # Already done:
-            return self._selected_set  # pylint: disable=access-member-before-definition
```
@bradenmacdonald, when I tested it, the memoization didn't really work and the condition was failing on every (re)load of the page containing the XBlock and the `make_selection` method was getting called. @kaizoku also saw the same behaviour and confirmed this.

The `make_selection` method call essentially behaves like a no-op whenever the number of blocks to be selected or the eligible blocks to be selected do not change. If either of those change, the memoization wouldn't help because the selection has to be made again.

bradenmacdonald replied:
@lgp171188 The memoization was only for a single request-response cycle, so it was expected that if you reload the page, it would call `make_selection` again. The main concern is just to make sure that a single load of the page results in no more than one call to `make_selection`.

lgp171188 replied:
> The memoization was only for a single request-response cycle,

@bradenmacdonald, during my testing, the `selected_children` method (which had the memoization and conditionally called the `make_selection` method) was called only once and that was the time of loading the page. So I am not able to understand how memoization will help. Did I miss something crucial to understanding this? Can you explain?

bradenmacdonald replied:
@lgp171188 In that case, it's fine :) I thought that there might be some circumstances where it got called more than once per page load, but if that's not the case, then there's no need to memoize it.

bradenmacdonald commented:
@lgp171188 @kaizoku  Can you please also check if the related block transformer needs to be updated? I believe it probably does.

https://github.com/edx/edx-platform/blob/8d34268948f7da732cda0813bebb15a3b514b76f/lms/djangoapps/course_blocks/transformers/library_content.py#L70-L134

If you're not familiar with it, this is the code that is used when the course structure is loaded via the course blocks API (such as when browsing the course from the mobile app), and it has separate logic for figuring out the child content of a library_content block; it must match the "normal" logic.

lgp171188 commented:
jenkins run bokchoy

lgp171188 commented:
@kaizoku, @bradenmacdonald, the bokchoy tests are repeatedly failing now due to unrelated LMS and studio help page test errors in `acceptance.tests.studio.test_studio_help` and `acceptance.tests.lms.test_lms_help` modules. I am not sure what can be done here. I will trigger another run of the bokchoy tests now.

Since this PR branch now has a conflict and is outdated, is it okay to rebase it on the master branch again?

lgp171188 commented:
jenkins run bokchoy

bradenmacdonald commented:
@lgp171188 Yep, please go ahead and rebase. That might also fix the bokchoy issue.

lgp171188 commented:
@bradenmacdonald the rebase seems to have fixed all the unrelated bokchoy errors. Do you have any review comments on the latest https://github.com/edx/edx-platform/pull/18675/commits/20959e450b5319519e51106c427c4d4ff5156138 commit having the changes to the course blocks API?

bradenmacdonald commented:
@lgp171188 Once @kaizoku has given the latest change(s) since his last approval a thorough review and signed off, I will review+test+merge this PR for you.

@mduboseedx does that sound good?

mduboseedx commented:
@bradenmacdonald Yes that sounds good. Thanks for your help!

kaizoku commented:
The new updates look good by me @lgp171188.
:+1:

- [✓] I tested this: (deployed in devstack; tested API endpoints which utilize the transformers, and regular XBlock views in a test course; ran applicable unit tests)
- [✓] I read through the code
- [n/a] ~~I checked for accessibility issues~~
- [n/a] ~~Includes documentation~~

bradenmacdonald commented:
Note: I am holding off on a final review+merge here while we investigate a report of a potential bug (or rather a situation where this fix may not be working).

bradenmacdonald commented:
@lgp171188 Is there a sandbox for this PR, and if so, is it up to date?

lgp171188 commented:
@bradenmacdonald, I have updated the sandbox details in the PR description and it has the latest changes from the source branch of this PR.

bradenmacdonald commented:
@lgp171188 @kaizoku The code looks good, but when I tested this on the sandbox, I found that the block transformer is not applying the randomization correctly.

Compare https://pr18675.sandbox.opencraft.hosting/courses/course-v1:BradenX+RAND+TEST/courseware/ae707bd895c94f52becc6ca8123ceb9f/6663816c6e3d4033a13ff78a898a3a25/1 with https://pr18675.sandbox.opencraft.hosting/api/courses/v1/blocks/?course_id=course-v1:BradenX%2BRAND%2BTEST&username=honor&depth=all&requested_fields=children (login as `honor`). In the LMS, the blocks are in the order 1,3,2 and 3,2,1 but in the course blocks API the `children` both show the order 1,2,3.

I believe the reason is that in the transformer `lms/djangoapps/course_blocks/transformers/library_content.py` the shuffle is applied to `state_dict['selected']` which is fine, but the actual order of the `children` returned by the API comes from some other part of the code (where?) which isn't reading that shuffled list correctly.

lgp171188 commented:
@bradenmacdonald, I will check the code to find out where this is issue is happening.

bradenmacdonald commented:
@lgp171188 Yes, you can ask @nasthagiri directly :)

@nasthagiri We'd appreciate a quick check from you if the above approach to modifying these block transformers to be able to specify the (new) order of their child blocks makes sense.

UmanShahzad commented:
@nasthagiri Were you able to check this one out?

nasthagiri commented:
@UmanShahzad @lgp171188 Sorry fellas - this PR got lost in my inbox.  Yes, this change is fine from my perspective.

lgp171188 commented on lms/djangoapps/course_blocks/transformers/tests/test_library_content.py:
```diff
@@ -165,3 +166,136 @@ def test_content_library(self):
+                self.transformers,
+            )
+            self.assertEqual(
+                set(trans_block_structure.get_block_keys()),
```
@kaizoku, thanks for catching this. The order is lost on conversion to a set. I have updated the PR and fixed this now.

natabene commented:
@lgp171188  Great, thanks. I am queueing this for engineering review, though it might take a while.

natabene commented:
@ormsbee Could you give this a look?

lgp171188 commented:
@ormsbee, @natabene, I have rebased the PR on top of the latest changes in the master branch and resolved the merge conflicts which happened with some recent changes.

natabene commented:
@lgp171188 Thanks! @ormsbee  Looks like this is ready for you when you have a chance.

lgp171188 commented:
@ormsbee, did you get a chance to check this?

lgp171188 commented:
@ormsbee, did you get a chance to check this?

natabene commented:
@lgp171188 I will see what we can do and if we can review it sooner.

lgp171188 commented:
jenkins run python

lgp171188 commented:
> Please squash and give a complete commit message with the context from the PR, and I'll merge. Thank you.

@ormsbee, thanks! I have rebased the PR branch, squashed the commits and updated the commit message explaining the context from this PR.

ormsbee commented:
jenkins run bokchoy

edx-webhook commented:
@lgp171188 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/edx-platform&entry.867055334=https://github.com/edx/edx-platform/pull/18675&entry.1484655318=https://github.com/lgp171188&entry.752974735=2018-07-30+12:55&entry.1917517419=2019-05-07+03:27&entry.2133058324=Yes) so we can improve your experience in the future.

ormsbee commented on lms/djangoapps/course_blocks/transformers/library_content.py:
```diff
@@ -176,3 +178,48 @@ def publish_event(event_name, result, **kwargs):
+            state_dict = get_student_module_as_dict(usage_info.user, usage_info.course_key, block_key)
+            library_children = block_structure.get_children(block_key)
+            ordering_data = {block[1]: position for position, block in enumerate(state_dict['selected'])}
+            library_children.sort(key=lambda block, data=ordering_data: data[block.block_id])
```
`lms.djangoapps.grades.tasks.recalculate_subsection_grade_v3` was failing on this line with the following stack trace:

```
exceptions:KeyError: u'23e9f816a170b4eb6422'
Traceback (most recent call last):
File "/edx/app/edxapp/edx-platform/manage.py", line 123, in <module>
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 364, in execute_from_command_line
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/django/core/management/__init__.py", line 356, in execute
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/djcelery/management/commands/celery.py", line 21, in run_from_argv
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/celery.py", line 793, in execute_from_commandline
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/base.py", line 311, in execute_from_commandline
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/celery.py", line 785, in handle_argv
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/celery.py", line 717, in execute
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/worker.py", line 179, in run_from_argv
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/base.py", line 274, in __call__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bin/worker.py", line 212, in run
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/worker/__init__.py", line 206, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bootsteps.py", line 123, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/bootsteps.py", line 374, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/base.py", line 131, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/prefork.py", line 119, in on_start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/asynpool.py", line 401, in __init__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 972, in __init__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/concurrency/asynpool.py", line 415, in _create_worker_process
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 1068, in _create_worker_process
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/process.py", line 137, in start
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/forking.py", line 105, in __init__
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/process.py", line 292, in _bootstrap
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 295, in run
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/billiard/pool.py", line 367, in workloop
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 349, in _fast_trace_task
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 240, in trace_task
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/newrelic/hooks/application_celery.py", line 85, in wrapper
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/trace.py", line 438, in __protected_call__
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 179, in recalculate_subsection_grade_v3
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 248, in _recalculate_subsection_grade
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/app/task.py", line 684, in retry
File "/edx/app/edxapp/venvs/edxapp/local/lib/python2.7/site-packages/celery/utils/__init__.py", line 248, in maybe_reraise
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 239, in _recalculate_subsection_grade
File "/edx/app/edxapp/edx-platform/lms/djangoapps/grades/tasks.py", line 310, in _update_subsection_grades
File "/edx/app/edxapp/edx-platform/lms/djangoapps/course_blocks/api.py", line 97, in get_course_blocks
File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content/block_structure/manager.py", line 76, in get_transformed
File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content/block_structure/transformers.py", line 116, in transform
File "/edx/app/edxapp/edx-platform/openedx/core/djangoapps/content/block_structure/transformers.py", line 153, in _transform_without_filters
File "/edx/app/edxapp/edx-platform/lms/djangoapps/course_blocks/transformers/library_content.py", line 225, in transform
File "/edx/app/edxapp/edx-platform/lms/djangoapps/course_blocks/transformers/library_content.py", line 225, in <lambda>
```

ormsbee replied:
@lgp171188 ^

lgp171188 replied:
@ormsbee, sorry for causing this issue. I will check this and try to reproduce it in my devstack.

This PR introduces a new `BlockStructureTransformer` sub-class `ContentLibraryOrderTransformer` to be run immediately after the `ContentLibraryTransformer`. The latter is a `FilteringTransformerMixin` sub-class and can only filter the children of a randomized content block and cannot randomize the order of selection.

This PR adds a change to the `ContentLibraryTransformer` to randomize/shuffle the order of the selected blocks before storing it in the `StudentModule`. The existing behaviour of making fresh selections only when the type of blocks to be chosen or the number of blocks to be chose has changed is retained as-is.

The new `ContentLibraryOrderTransformer` then tries arrange the children blocks based on the stored order using the sort method above which is throwing an exception in the task.

On taking a quick look at the exception and the code, it appears that this can happen only when the selected blocks stored in the `StudentModule` no longer match the ones that are actually available in the current block structure. And this appears to be a weird and unexpected error because the `ContentLibraryTransformer`, which makes the selections and stores it in the `UserModule` and the `ContentLibraryOrderTransformer` which uses it, are always run together one after the other.

Do you see any obvious issues with the approach or have any pointers on what might have gone wrong? That will be helpful because my understanding of the block structure and its transformation is limited and I could have missed something crucial while working on this.

That said, I will still try to debug and investigate this and come up with a fix. Thanks!

lgp171188 replied:
It might be helpful to debug this issue further if there are some more details about the course/course structure are available so that I can try to reproduce the issue.
